docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,628 @@
+
+This step will run coercions and validations defined in the provided block. If the underlying contract isn’t valid (at least one validation failed), then the execution flow will stop.
+
+### `transaction`
+
+This step is a bit special, as it will wrap any other steps defined in its block inside a SQL transaction.
+
+### `options`
+
+This step is another special one, as it’s similar to a contract (without the validations part), but for options your service can take. This is useful if you need to change your service behavior depending on certain conditions. Also, that step can’t fail.
+
+### Steps arguments
+
+Each step is called with the service context. To access a value in it, just provide its key as a keyword argument.
+
+## The context object
+
+The only state a service maintain is its context object. This is where each step can put data to be used by other steps. Most of the time, you don’t need to access the context directly, as specialized steps (such as `params`, `model` or `options`) will store the proper data for you. But sometimes it’s necessary (even if uncommon). In those cases, it’s just a matter of using the context like you would with a hash:
+```ruby
+def first_step
+  context[:my_special_key] = "My special value"
+end
+
+…
+
+# Then in a later step, you can use it like any other key from the context
+def another_step(my_special_key:)
+  # do something with `my_special_key`
+end
+```
+
+## Handling service results
+
+Once a service has been called, it will return a result object, to know whether the call was a success. In the case of a failure, the result object can be inspected to know what failed and why.
+
+Each step will store its outcome in the result object, accessible through special keys (like `result.model.user` for example). While this is nice, this would be tedious to manually check the result object. That’s why there’s a built-in feature allowing to run the service, match step outcomes and act upon results using a custom DSL.
+
+This feature makes the use of a service a breeze. Continuing with our `User::UpdateUsername` service, this is how we could use it inside a controller (but it can work anywhere, not just in controllers):
+```ruby
+def update
+  User::UpdateUsername.call(service_params) do |result|
+    on_success { |user:| render(json: success_json.merge(new_username: user.username)) }
+    on_failure { render(json: failed_json, status: 422) }
+    on_failed_contract { |contract| render(json: failed_json.merge(errors: contract.errors.full_messages), status: 400) }
+    on_model_not_found(:user) { raise Discourse::NotFound }
+    on_failed_policy(:can_update_username) { raise Discourse::InvalidAccess.new }
+  end
+end
+```
+
+Passing a block to `.call` allows to “match” an outcome, a bit like if we were using pattern matching. There are two generic matchers (`on_success` and `on_failure`) and each specialized step has at least one dedicated matcher. The complete detailed list is available in the API section.
```

Comment:
If I understand correctly, `on_failure` will only be used if there is no specific block defined for the failure? Should we call that out in this paragraph?

My initial thought was "I guess on_failure is called for all failures **IN ADDITION TO** the specialised failures". But of course, that wouldn't make sense with the multiple `render: ` we have in this example. 👌

Reply:
I added this:

> 💡 `on_failure` is like a catch-all rule, it will match only if the service fails _and_ no other more specialized matcher matches.

Do you think it reads well enough?

Reply:
LGTM :+1: